design/ceph/multus-network.md

```diff
@@ -78,7 +78,78 @@ Nothing to do in particular since they don't use any service IPs.
 
 ### CSI pods
 
-We can add annotations to these pods and they can reach out the Ceph public network, then the driver will expose the block or the filesystem normally.
+Steps must be taken to fix a CSI-with-multus issue documented
+[here](https://github.com/rook/rook/issues/8085). To summarize the issue:
+when a CephFS/RBD volume is mounted in a Pod using cephcsi and then the CSI CephFS/RBD plugin is
+restarted or terminated (e.g. by restarting or deleting its DaemonSet), all operations on the volume
+become blocked, even after restarting the CSI pods. The only workaround is to restart the node where
+the cephcsi plugin pod was restarted.
+
+When deploying a CephCluster resource configured to use multus networks, a multus-connected network
+interface will be added to the host network namespace for all nodes that will run CSI plugin pods.
+
+The Rook-Ceph Operator's CSI controller creates a DaemonSet configured to use the
+`network.selectors.public` network specified for the CephCluster. This DaemonSet runs on all the
+nodes that will have CSI pods. Its Pods exist to "hold" a particular network interface that the CSI
+pods can reliably connect to. These Pods will be referred to as the "holder" Pods.
+
+These Pods should only be stopped and restarted when a node is stopped so that volume operations do
+not become blocked. The Rook-Ceph Operator's CSI controller should set the DaemonSet's update
+strategy to `OnDelete` so that the pods do not get deleted if the DaemonSet is updated while also
+ensuring that the pods will be updated on the next node reboot.
+
+Furthermore, a new Rook-specific CSI sidecar will run in CSI plugin Pods alongside the Ceph CSI
+plugin process. This sidecar will watch for holder Pods, and when a holder Pod is created or
+updated, it will move the public network from the Pod's network namespace to the host node's network
+namespace. In this way, even if the CSI pods are restarted, the network interface will not change,
+and the operations to the Ceph volume will not be blocked.
+
+If a holder Pod is deleted, the interface hold will be lost, so the sidecar should remove the
```

Comment:
>The CSI pods will all run on the host network in this design, so I believe the nodeplugin IPs should remain the same. However, the IPs used to connect to RBD/CephFS volumes will change due to getting a new multus lease. I'm not sure if that affects the DR cases you mention or not.

This is what concerns me. if we blocklist the IP and expect the blocklisting to remain until we remove it, as the IP used to connect to RBD/CephFS volumes will change so the connection will be restored back and it will cause the problem as pods will get read/write access. currently, multus is not planned for Metro DR am just making sure if it's used in the future this should not cause any problems. (even if someone designs a DR feature with multus and Rook managing an external ceph cluster things might not work?) cc @raghavendra-talur 

> I'm happy to coordinate with you to make sure this isn't going to break anything for DR.

Thanks, @BlaineEXE I can join the Rook meeting and provide more details about the problem am seeing.

Reply:
I have two thoughts on this:

I think what you are describing is a challenge with doing IP blocklisting in a Kubernetes environment. Any pod might get killed and restart with a new IP for a lot of reasons. I don't think fencing via IP is a particularly robust strategy in a K8s environment.

I'm not sure it matters in this case. If a holder pod is removed, it will have the same effect as the csi plugin pod restarting today: user pods on the node will lose I/O to RBD/CephFS volumes on the node because the interface has changed. I don't believe there is a way to re-establish an I/O connection to the ceph cluster once that IP changes. At least, that is my understanding, and I believe that is the underlying reason for implementing this multus fix in the first place. 

If ther **is** a way to re-establish an I/O connection to the ceph cluster once the IP changes, then I think we should abandon this work and implement the re-connection logic in the CSI driver so that we don't have to create this hacky workaround for the multus case.